enhancement: Redshift

MattiasMTS opened the pull request:
# 👷 What has changed 👷 
The focus of this PR is to open up this awesome plugin for Redshift users! 

**I'm opening this PR as a draft for early review. I'll push some more changes with respect to tests, documentation, etc**

--- 
### Redshift

- Added a new client with a similar structure as the Postgres client but with some differences. Refactored the Layout schema for both clients to re-usable function.
- Extended the layout to make use of `VIEWS`
- Added more code actions to action 1 for both `VIEWS` and `TABLES`
- Made upper case letters for keywords in helpers

### Layout
- added sorting with respect to history to get the latest query at the top and alphabetical on the schema

### Logging
Extended the logging library with formatted functions and refactored a few use cases where this was needed.

### Bash
I've `shellfmt` as formatter on nvim and I was in `build.sh`, which led to the formatted version. Let me know if you want me to revert this. Perhaps it is more worth adding a `pre-commit-config.yaml` to fix this for future stuff. We can add more configs with respect to `sh`, `go`, `lua`, etc.

kndndrj commented:
Hey @MattiasMTS thanks for the contribution! 

The changes look nice - don't worry about fomatters (for bash I used something different - `beutysh` or something like that,  but it doesn't matter)

The only thing which I'd like to ask you is that I just prepared a PR for database switching, so I'd like to ask you to revisit this PR once that is merged (I'll try to ping you here) It shouldn't be too complicated, but I expect a few merge confilicts.

MattiasMTS commented:
> Hey @MattiasMTS thanks for the contribution! 
> 
> 
> 
> The changes look nice - don't worry about fomatters (for bash I used something different - `beutysh` or something like that,  but it doesn't matter)
> 
> 
> 
> The only thing which I'd like to ask you is that I just prepared a PR for database switching, so I'd like to ask you to revisit this PR once that is merged (I'll try to ping you here) It shouldn't be too complicated, but I expect a few merge confilicts.

Sounds great! I had a gut feeling this was coming up, hence the draft PR. 

I'll hold and rebase after you've merged database switching. Until then I'll pick up something in the backlog/issue. 

Thanks for fast response! :)

kndndrj commented:
Here it is. I merged into master.

> Until then I'll pick up something in the backlog/issue
whoa, nice :)

Also feel free to ask questions, I'd be happy to help :)

MattiasMTS commented:
> Here it is. I merged into master.
> 
> > Until then I'll pick up something in the backlog/issue
> 
> whoa, nice :)
> 
> Also feel free to ask questions, I'd be happy to help :)

Sweet. I'll take a peak and rebase. 😋

MattiasMTS commented:
This should now be ready for "real review". Thanks for the initial feedback @kndndrj ! I addressed the rebasing issues, main thing was the new list helper method which in this case needed to accept the vars argument that the helper:get uses as well in order to list the option difference for e.g. VIEW, TABLE, etc. Perhaps worth to add e.g. "Functions" here in the future.

--- 

#### `client.go`

**NB**  I did some extension of the `client.go` file, which is to add the interfaces `DatabaseClient` and `DatabaseConnetion`. Also updated the return statement of some methods to return interfaces rather than the concrete result. It isn't idiomatic that methods return interfaces rather than the concrete stuff but this refactoring facilitated testing.

The interfaces were needed in order to add mock structs, see how I did it in the `redshift_test.go` file. Maybe this isn't the prettiest way of doing it. Let me know if you have any other ideas on how to do this and I'm more than happy to do it :)

#### `history.go` & `result.go` 

Added `SetCallBack` and `SetCustomHeader` interface methods to. Updated the `HistoryRows` struct with placeholder method to fulfil the interface of `models.IterResult`. These were needed due to the return statement of the results now returning interfaces rather than the concrete stuff. Again, this made testing easier. Perhaps not the best way to do it since large interfaces become difficult to handle in the long run.. :/

kndndrj commented:
I did a quick review from my phone, and overall I like the changes - reminds me of some stuff that I forgot to clean up myself. I'd still like to take a closer look when I have some time though.

Thanks for the contribution and a detailed overview :)

MattiasMTS commented:
Looking forward to it!

Thanks for awesome response and feedback during these PRs 😄

MattiasMTS commented:
I'll add this as a reminder to myself to actually reconsider using the redshiftdata api https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/redshiftdata rather than using the `database/sql` package.

kndndrj commented on ci/build.sh:
```diff
@@ -4,37 +4,44 @@
 set -e
```
Nice of you that you want to format the script, but it was formatter with a different formatter already 😅 

I think we can remove these changes if you agree.

MattiasMTS replied:
Deleted the commit. Totally agree. Is it worth adding some pre-commit-config in the future? 😋

kndndrj replied:
It would make sense, yeah. although I'm not the most familiar with those configs. If you want, you can come up with something

kndndrj commented on dbee/clients/common/client.go:
```diff
@@ -8,7 +8,20 @@ import (
 )
 
-// default sql client used by other specific implementations
+// DatabaseClient is an interface that represents a database client.
+type DatabaseClient interface {
+	Conn() (DatabaseConnection, error)
+	Close()
+	Swap(db *sql.DB)
+}
+
+// DatabaseConnection is an interface that represents a database connection.
+type DatabaseConnection interface {
+	Close() error
+	Exec(query string) (models.IterResult, error)
+	Query(query string) (models.IterResult, error)
+}
+
 type Client struct {
```
The point of these `common` files is to provide a result builder and a client that works for "most" drivers. If the driver's functionality is too specific, i think it's better that it uses it's own implementation.

kndndrj commented on dbee/clients/common/client.go:
```diff
@@ -19,12 +32,14 @@ func NewClient(db *sql.DB) *Client {
 }
 
-func (c *Client) Conn() (*Conn, error) {
+func (c *Client) Conn() (DatabaseConnection, error) {
```
Tying to the comment above, we want THE Connection, not A connection here.

Same for other places

MattiasMTS replied:
[b9d0035](https://github.com/kndndrj/nvim-dbee/pull/34/commits/b9d0035a1eeda0d6b564928c154cd39e1d7ea90e)

kndndrj commented on dbee/clients/redshift.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// fetchPsqlLayouts fetches the layout from the postgres database.
+func fetchPsqlLayouts(rows models.IterResult, dbType string) ([]models.Layout, error) {
```
Nice to repurpose the code, but doesn't it make more sense to have this function in `postgres.go`? :D

Also, I like `Postgres` or at least `PG` more that `Psql` :P

MattiasMTS replied:
[c1b18cb](https://github.com/kndndrj/nvim-dbee/pull/34/commits/c1b18cb5b51a3da627fe062117306e1836fc1478)

kndndrj commented on dbee/clients/redshift.go:
```diff
@@ -0,0 +1,163 @@
+// done here. The ordering is done in the lua frontend.
+func (c *RedshiftClient) Layout() ([]models.Layout, error) {
+	query := `
+	SELECT
+    trim(n.nspname) AS schema_name,
+    trim(c.relname) AS table_name,
+    CASE
+        WHEN c.relkind = 'v' THEN 'VIEW'
+        ELSE 'TABLE'
+    END AS table_type
+FROM
+    pg_class AS c
+JOIN
+    pg_namespace AS n ON c.relnamespace = n.oid
+WHERE
+    n.nspname NOT IN ('information_schema', 'pg_catalog');
+`
```
even though it's a raw string, i think we can handle some nice indentation

MattiasMTS replied:
[120c3f2](https://github.com/kndndrj/nvim-dbee/pull/34/commits/120c3f2968608fcc67886b1eae6a8108dafce466)

kndndrj commented on dbee/clients/redshift_test.go:
```diff
@@ -0,0 +1,477 @@
+package clients
```
Note for future self: I'd like to add a more common way of mocking connections and rows in the future. Also integration tests would be nice :)

MattiasMTS replied:
[d91671b](https://github.com/kndndrj/nvim-dbee/pull/34/commits/d91671b66fbae0fb20aeeb25f994c4def457d9fd)

I agree. I removed the tests now and will probably add them back in the future once it is easier to mock out connections/rows from the common package 👍

kndndrj commented on dbee/clients/common/client.go:
```diff
@@ -124,9 +134,6 @@ func (c *Conn) Query(query string) (*Result, error) {
 				val := *columnPointers[i].(*any)
-				// TODO: this breaks some types with some drivers (namely sqlserver newid()):
-				// add a generic way of doing this with ResultBuilder
-				// fix for some strings being interpreted as bytes
```
please don't remove those TODO's :(

MattiasMTS replied:
[02ed575](https://github.com/kndndrj/nvim-dbee/pull/34/commits/02ed575a21dc72a5b39ccf16ad4a77b402dcbd11)

kndndrj commented on dbee/main.go:
```diff
@@ -53,32 +54,31 @@ func main() {
 		// This must be called before bufferOutput is used
 		p.HandleFunction(&plugin.FunctionOptions{Name: "Dbee_set_results_buf"},
-			func(v *nvim.Nvim, args []int) error {
+			func(_ *nvim.Nvim, args []int) error {
```
```suggestion
			func(args []int) error {
```

kndndrj commented on dbee/models/layout.go:
```diff
@@ -36,6 +37,8 @@ func (s LayoutType) String() string {
 		return "history"
 	case LayoutTypeDatabaseSwitch:
 		return "database_switch"
+	case LayoutTypeView:
```
I was thinking about that a while ago, and I couldn't make up my mind.
It looks like we have views as a separate layout type now :D

kndndrj commented on dbee/models/result.go:
```diff
@@ -34,6 +34,8 @@ type (
 		Header() (Header, error)
 		Next() (Row, error)
 		Close()
+		SetCallback(callback func())
+		SetCustomHeader(header Header)
```
I don't want this here, because these two methods aren't required to make a functioning `IterResult`

They are just helper methods in the `common.Result` builder for creating specific implementations of this interface.

MattiasMTS replied:
[b9d0035](https://github.com/kndndrj/nvim-dbee/pull/34/commits/b9d0035a1eeda0d6b564928c154cd39e1d7ea90e)

kndndrj commented on dbee/vim/logger.go:
```diff
@@ -92,14 +93,30 @@ func (l *Logger) Debug(msg string) {
 	l.log(nvim.LogDebugLevel, msg)
 }
 
+func (l *Logger) Debugf(format string, args ...any) {
+	l.log(nvim.LogDebugLevel, fmt.Sprintf(format, args...))
+}
+
```
very nice, thanks :)

kndndrj commented on lua/dbee/config.lua:
```diff
@@ -75,14 +75,17 @@ M.default = {
       history = {
         icon = "",
         icon_highlight = "Constant",
+        text_highlight = "",
```
I see you got the annoying sumneko update too 😅

kndndrj commented on lua/dbee/handler/init.lua:
```diff
@@ -169,6 +169,7 @@ end
 function Handler:layout()
   -- in case there are no sources defined, return a helper layout
   if #self.lookup:get_sources() < 1 then
+    vim.print(self:layout_help())
```
```suggestion
```

kndndrj commented on lua/dbee/handler/conn.lua:
```diff
@@ -218,10 +218,21 @@ function Conn:layout()
-    table.sort(layout_go, function(k1, k2)
-      return k1.type .. k1.name < k2.type .. k2.name
-    end)
+    -- sort children of keys in layout
+    for _, lgo in ipairs(layout_go) do
+      if lgo.name == "history" then
+        -- sort descendingly, latest query in the top
+        -- due to table.name is epoch time.
+        table.sort(lgo.children, function(k1, k2)
+          return k1.name > k2.name
+        end)
+      elseif lgo.name == "structure" then
+        -- sort ascendingly, alphabetically
+        table.sort(lgo.children, function(k1, k2)
+          return k1.name < k2.name
+        end)
+      end
+    end
```
I wonder if it's really more intuitive to have more recent history near the bottom?

Also, the `.type .. .name` sorts nicer in my opinion, especially if you have views and tables in one "folder" for example

MattiasMTS replied:
`.type .. .name` is definitely worth to group the materialization + name separately; nice feedback!

Regarding the recent history, IMHO, imagine we are growing the amount of results/queries to +1000 and I want to easy go back to my previous execution I'd have to go to the bottom of the history bar to see it, rather than just the top one (which is always at the top) and allow you to easy see it again.

Should probably have some sort of "limit" of the historical view or the ability to delete them if we want - just with the motivation on how the UI tree would look if the amount of queries is growing large 😅

MattiasMTS replied:
[36bea65](https://github.com/kndndrj/nvim-dbee/pull/34/commits/36bea65dcebedcaf9d14c9ec0b3875fb0fea9c22)

The sorting in `History` is still kept the same as the latest result at the top.

The sorting in `Structure` is now done by 
1. alphabetically descending based on the `schemas`
2. Within each schema we sort:
    1. First by the name of each materilization
    2. The name of the materialization (i.e. `VIEW` , `TABLE`, `FUNCTION`, etc...)

kndndrj replied:
Nice, didn't think of that. Well, I think my judgement is as good as yours (or worse), so I trust your taste on this one :)

kndndrj reviewed (changes requested):
Thanks again for submitting this PR and sorry that it took me so long to do a propper review.

I'd be really thankful if you address the changes, so we can merge this cool feature!

kndndrj commented on dbee/conn/history.go:
```diff
@@ -393,4 +394,13 @@ func (r *HistoryRows) Next() (models.Row, error) {
+}
+
+func (r *HistoryRows) SetCustomHeader(header models.Header) {
+	// no-op
```
Are these methods an artifact of the deleted interface?

MattiasMTS replied:
Indeed they are [c82b85e](https://github.com/kndndrj/nvim-dbee/pull/34/commits/c82b85e230b61eec0664f2e0e74e23decc516d29)

kndndrj commented on lua/dbee/handler/conn.lua:
```diff
@@ -218,10 +218,35 @@ function Conn:layout()
+    ---@param children _LayoutGo[] -- children to sort
+    ---@param field string -- field to sort by
+    ---@param ascending bool -- set to true to sort ascendingly
+    local function sortChildrenByField(children, field, ascending)
```
cool, just snake case please 😅

MattiasMTS replied:
[15e5262](https://github.com/kndndrj/nvim-dbee/pull/34/commits/15e5262092cb45069eafcda260e01735217d1bf3) sorry, too much scala

kndndrj reviewed (commented):
Except for the 2 minor nitpicks, I think the PR looks fine now.
I'll do a quick test and merge it today or tomorrow. Thanks again!

kndndrj commented:
thanks for the contribution. I tested some stuff and fixed what didn't work and added some extra stuff.
